hypershift-operator/controllers/awsendpointservice/awsendpointservice_controller.go

```diff
@@ -0,0 +1,208 @@
+	if !controllerutil.ContainsFinalizer(awsEndpointService, finalizer) {
+		controllerutil.AddFinalizer(awsEndpointService, finalizer)
+		if err := r.Update(ctx, awsEndpointService); err != nil {
+			if apierrors.IsConflict(err) {
+				return ctrl.Result{Requeue: true}, nil
+			}
+			return ctrl.Result{}, fmt.Errorf("failed to add finalizer: %w", err)
+		}
+	}
+
+	// Reconcile the AWSEndpointService
+	err = reconcileAWSEndpointService(ctx, awsEndpointService, r.ec2Client, r.elbv2Client)
+	if err != nil {
+		meta.SetStatusCondition(&awsEndpointService.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.AWSEndpointServiceAvailable),
+			Status:  metav1.ConditionFalse,
+			Reason:  hyperv1.AWSErrorReason,
+			Message: err.Error(),
+		})
+	} else {
+		meta.SetStatusCondition(&awsEndpointService.Status.Conditions, metav1.Condition{
+			Type:    string(hyperv1.AWSEndpointServiceAvailable),
+			Status:  metav1.ConditionTrue,
+			Reason:  hyperv1.AWSSuccessReason,
+			Message: "",
+		})
+	}
+	r.Status().Update(ctx, awsEndpointService)
+	return ctrl.Result{}, err
+}
+
+func reconcileAWSEndpointService(ctx context.Context, awsEndpointService *hyperv1.AWSEndpointService, ec2Client ec2iface.EC2API, elbv2Client elbv2iface.ELBV2API) error {
+	log := logr.FromContext(ctx)
+
+	serviceName := awsEndpointService.Status.EndpointServiceName
+	if len(serviceName) != 0 {
+		// check if Endpoint Service exists in AWS
+		output, err := ec2Client.DescribeVpcEndpointServiceConfigurationsWithContext(ctx, &ec2.DescribeVpcEndpointServiceConfigurationsInput{
+			Filters: []*ec2.Filter{
+				{
+					Name:   aws.String("service-name"),
+					Values: []*string{aws.String(serviceName)},
+				},
+			},
+		})
+		if err != nil {
+			return err
+		}
+		if len(output.ServiceConfigurations) == 0 {
+			// clear the EndpointServiceName so a new Endpoint Service is created on the requeue
+			awsEndpointService.Status.EndpointServiceName = ""
+			return fmt.Errorf("endpoint service %s not found", serviceName)
+		}
+		return nil
+	}
+
+	// determine the LB ARN
+	lbName := awsEndpointService.Spec.NetworkLoadBalancerName
+	output, err := elbv2Client.DescribeLoadBalancersWithContext(ctx, &elbv2.DescribeLoadBalancersInput{
+		Names: []*string{aws.String(lbName)},
+	})
+	if err != nil {
+		return err
+	}
+	if len(output.LoadBalancers) == 0 {
+		return fmt.Errorf("NLB %s not found", lbName)
+	}
+	lbARN := output.LoadBalancers[0].LoadBalancerArn
+
+	// create the Endpoint Service
+	createEndpointServiceOutput, err := ec2Client.CreateVpcEndpointServiceConfigurationWithContext(ctx, &ec2.CreateVpcEndpointServiceConfigurationInput{
+		AcceptanceRequired:      aws.Bool(false),
+		NetworkLoadBalancerArns: []*string{lbARN},
+	})
+	if err != nil {
+		return err
+	}
+	serviceName = *createEndpointServiceOutput.ServiceConfiguration.ServiceName
+	log.Info("endpoint service created", "serviceName", serviceName)
+	awsEndpointService.Status.EndpointServiceName = serviceName
```

Comment:
The AWS create call and status update are not atomic — if the create succeeds but the status update fails, won't subsequent create calls fail due to an "already exists" type error? If so does this need a check to no-op if create returns an already-exists error?

Reply:
Yes, this is true and I have seen this. Probably should handle that is some way.  Not sure what the best way is.

On the second reconcile (requeued after the first fails on the status update), the AWS error is something like "LB is already associated with an Endpoint Service", but it doesn't tell you which one so you don't know which to adopt i.e. just set `status.endpointServiceName` to the existing Endpoint Service.

Best solution I can think of is, if `CreateVpcEndpointServiceConfigurationWithContext` fails with that particular error, list all Endpoint Services, find the one with our NLB, and adopt it.

Reply:
>Best solution I can think of is, if CreateVpcEndpointServiceConfigurationWithContext fails with that particular error, list all Endpoint Services, find the one with our NLB, and adopt it.

I don't have a better idea, but if you already observed this during your own small scale testing, does it seem safe to assume it'll happen in the wild? If so maybe it's worth trying to address it now... your adoption idea makes sense to me, @csrwng any thoughts?

Reply:
Key questions I can think of are how reliable is the adoption algo and what's the impact of accidentally adopting the wrong thing?

Reply:
i.e. depending on impact of an incorrect adoption maybe a failure and do-over is better than risking a wrong guess

Reply:
there is a 1:1 relationship NLB to Endpoint Service and a 1:N relationship Endpoint Service to NLBs.  the adoption should be safe.

Reply:
for the record the exact error is something like `InvalidParameter: LBs are already associated with another VPC Endpoint Service Configuration: arn:aws:elasticloadbalancing:us-west-1:820196288204:loadbalancer/net/a1ef220129ff44c17a7416e0353614ce/419b61a779cf7399`